hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/AbstractQueueCapacityCalculator.java

```diff
@@ -0,0 +1,228 @@
+                                                 String label);
+
+  /**
+   * Executes all logic that must be called prior to the effective resource value calculations.
+   *
+   * @param resourceCalculationDriver driver that contains the parent queue on which the prerequisite
+   *                                  calculation should be made
+   */
+  public void calculateResourcePrerequisites(ResourceCalculationDriver resourceCalculationDriver) {
+    for (String label : resourceCalculationDriver.getParent().getConfiguredNodeLabels()) {
+      // We need to set normalized resource ratio only once per parent
+      if (resourceCalculationDriver.getNormalizedResourceRatios().isEmpty()) {
+        setNormalizedResourceRatio(resourceCalculationDriver, label);
+      }
+    }
+  }
+
+  /**
+   * Returns all resource names that are defined for the capacity type that is
+   * handled by the calculator.
+   *
+   * @param queue queue for which the capacity vector is defined
+   * @param label node label
+   * @return resource names
+   */
+  protected Set<String> getResourceNames(CSQueue queue, String label) {
+    return getResourceNames(queue, label, getCapacityType());
+  }
+
+  /**
+   * Returns all resource names that are defined for a capacity type.
+   *
+   * @param queue        queue for which the capacity vector is defined
+   * @param label        node label
+   * @param capacityType capacity type for which the resource names are defined
+   * @return resource names
+   */
+  protected Set<String> getResourceNames(CSQueue queue, String label,
+                                         ResourceUnitCapacityType capacityType) {
+    return queue.getConfiguredCapacityVector(label)
+        .getResourceNamesByCapacityType(capacityType);
+  }
+
+  /**
+   * Sets capacity and absolute capacity values based on minimum and maximum effective resources.
+   *
+   * @param clusterResource cluster resource for the corresponding label
+   * @param queue child queue for which the capacities are set
+   * @param label node label
+   */
+  public static void setQueueCapacities(
+      Resource clusterResource, CSQueue queue, String label) {
+    if (!(queue instanceof AbstractCSQueue)) {
+      return;
+    }
+
+    AbstractCSQueue csQueue = (AbstractCSQueue) queue;
+    ResourceCalculator resourceCalculator = csQueue.resourceCalculator;
+
+    CSQueue parent = queue.getParent();
+    if (parent == null) {
+      return;
+    }
+    // Update capacity with a float calculated from the parent's minResources
+    // and the recently changed queue minResources.
+    // capacity = effectiveMinResource / {parent's effectiveMinResource}
+    float result = resourceCalculator.divide(clusterResource,
+        queue.getQueueResourceQuotas().getEffectiveMinResource(label),
+        parent.getQueueResourceQuotas().getEffectiveMinResource(label));
+    queue.getQueueCapacities().setCapacity(label,
```

Comment:
I'm not thirlled by the idea of a calculator actually making changes of the queues. It is more like a driver's job. Calculator should calulate but don't write the result back to the queues.

Reply:
Done and this logic was moved into the handler.